Add MetricsEventSource

noahfalk opened the pull request:
The feature is still a work in progress but wanted to let others
see it in its current state while I am refining it.

Our out-of-process tools like dotnet-counters and dotnet-monitor need
access to the metrics produced by the new Meter APIs without
requiring the app to take any dependency on a separate OpenTelemetry
library. System.Diagnostics.Metrics EventSource is a new source designed
to let those tools access this data. The EventSource includes high
performance in-proc pre-aggregation capable of observing
millions of instrument invocations/sec/thread with low CPU overhead.

This change does not create any new BCL API surface, the aggregated
data is solely exposed by subscribing to the EventSource such as
using ETW, EventPipe, Lttng, or EventListener. For anyone wanting
in-process APIs to consume the data they could either use MeterListener
for unaggregated data or a library such as OpenTelemetry for
pre-aggregated data.

Todo list:
- [x] - I think we need a configuration on the EventSource to limit the max number of metrics to track. This ensures that even if someone goes nuts with tags we won't accidentally OOM the app trying to insert all those aggregations into the dictionaries.

noahfalk committed:
Add MetricsEventSource

The feature is still a work in progress but wanted to let others
see it in its current state while I am refining it.

Our out-of-process tools like dotnet-counters and dotnet-monitor need
access to the metrics produced by the new Meter APIs without
requiring the app to take any dependency on a separate OpenTelemetry
library. System.Diagnostics.Metrics EventSource is a new source designed
to let those tools access this data. The EventSource includes high
performance in-proc pre-aggregation capable of observing
millions of instrument invocations/sec/thread with low CPU overhead.

This change does not create any new BCL API surface, the aggregated
data is solely exposed by subscribing to the EventSource such as
using ETW, EventPipe, Lttng, or EventListener. For anyone wanting
in-process APIs to consume the data they could either use MeterListener
for unaggregated data or a library such as OpenTelemetry for
pre-aggregated data.

noahfalk commented:
@tarekgh @dotnet/dotnet-diag @cijothomas @reyang @victlu @wiktork @jander-msft @shirhatti 
I'm still cleaning this up but giving a heads up on what is coming. If you have any feedback on the broad functionality/design/structure I'd much appreciate it. Feedback on implementation details are welcome too but you may find it more efficient to let me clean it more and switch to non-draft PR before bothering to dig in at that level.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,274 @@
+        private List<Predicate<Instrument>> _instrumentConfigFuncs = new();
+        private TimeSpan _collectionPeriod;
+
+        private ConcurrentDictionary<Instrument, InstrumentState> _instrumentStates = new ConcurrentDictionary<Instrument, InstrumentState>();
```
Nit:
```suggestion
        private ConcurrentDictionary<Instrument, InstrumentState> _instrumentStates = new();
```
Other fields also use the short form.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,274 @@
+            if (_collectTask != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                Debug.Assert(false);
```
```suggestion
```
Is it necessary when the next line throws?
Or use `Debug.Fail`?

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,274 @@
+                throw new InvalidOperationException("CollectionPeriod must be >= 1 sec");
+            }
+
+            var token = _cts.Token;
```
```suggestion
            CancellationToken token = _cts.Token;
```
BTW: `token` is a quite generic name, can this be named `cancellationToken`?

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,274 @@
+            _cts.Cancel();
+            if (_collectTask != null)
+            {
+                _collectTask.Wait();
```
```suggestion
                _collectTask.GetAwaiter().GetResult();
```
? (to avoid the AggregateException).

Is DisposeAsync an option here?

noahfalk replied:
This got changed to thread.Join() as part of a switch from Task to Thread

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,274 @@
+            }
+            else if (genericDefType == typeof(Histogram<>))
+            {
+                return () => new ExponentialHistogramAggregator(new PercentileAggregation(new double[] { 50, 95, 99 }));
```
I assume the array is "constant", so move this to a static readonly field to avoid repeated allocations?

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Aggregator.cs:
```diff
@@ -0,0 +1,64 @@
+            _labels = labels;
+        }
+
+        public KeyValuePair<string, string>[] Labels => _labels;
```
Use auto-properties as in the other types here too?
Or make the fields readonly.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,478 @@
+            return GetAggregatorSlow(labels);
+        }
+
+        public TAggregator GetAggregatorSlow(ReadOnlySpan<KeyValuePair<string, object?>> labels)
```
```suggestion
        [MethodImpl(MethodImplOptions.NoInlining)]
        public TAggregator GetAggregatorSlow(ReadOnlySpan<KeyValuePair<string, object?>> labels)
```
To be sure that the JIT won't inline it (as slow path)?

noahfalk replied:
I decided to reduce the explicit inlining control rather than add more. Its entirely possible that it would help but since I am not planning to analyze the codegen across runtime versions and architectures it felt a little easier to let the JIT use its default heuristics for now. If we need to go questing for more nanoseconds in the future we could always give it a closer look then.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,478 @@
+                object? state = _stateUnion;
+                if (state == null)
+                {
+                    var newState = _createAggregatorFunc();
```
Use concrete type instead of `var` here.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,478 @@
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)LabelMany;
+            }
+            Debug.Assert(false);
```
Necessary?

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,108 @@
+        {
+            lock (this)
+            {
+                ref long bits = ref Unsafe.As<double, long>(ref measurement);
```
```suggestion
                long bits = BitConverter.DoubleToInt64Bits(measurement);
```

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,108 @@
+        // lower bound for positive valued buckets
+        private double GetBucketCanonicalValue(int exponent, int mantissa)
+        {
+            double result = 0;
+            ref long bits = ref Unsafe.As<double, long>(ref result);
+            bits = ((long)exponent << ExponentShift) | ((long)mantissa << _mantissaShift);
+            return result;
```
Is this swapped?
In L103 `bits` is `0` as `result = 0` in L102, then `bits` is assigned in L104 to a value. `result` at return on L105 is still `0`.

```suggestion
            long bits = ((long)exponent << ExponentShift) | ((long)mantissa << _mantissaShift);
            return BitConverter.Int64BitsToDouble(bits);
```
(but I'm not sure if I understood the intention here correct).

noahfalk replied:
The ref is like creating a pointer in C so any write to bits also changes the value of result. However the BitConverter approach you suggested is more clear so I'm going with that instead👍

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentAggregation.cs:
```diff
@@ -0,0 +1,53 @@
+
+    internal static class MeasurementAggregations
+    {
+        public static RateAggregation Rate = new RateAggregation();
```
Can these be `static readonly`?
And use the short form `new()` as above?

noahfalk replied:
I deleted them during some refactoring (but if I had needed to keep them then yep, readonly would have worked)

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,305 @@
+    [EventSource(Name = "System.Diagnostics.Metrics")]
+    internal class MetricsEventSource : EventSource
+    {
+        public static MetricsEventSource Logger = new MetricsEventSource();
```
```suggestion
        public static readonly MetricsEventSource Logger = new();
```

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,305 @@
+
+            private string FormatTags(KeyValuePair<string, string>[] labels)
+            {
+                StringBuilder sb = new StringBuilder();
```
Should this be `ValueStringBuilder` to avoid the allocation(s) or doesn't it matter here?

noahfalk replied:
I expect it not to matter. This code path runs each collection interval which is typically 1 sec - 10 minutes in length. Given the infrequency I biased towards simple code.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,305 @@
+
+            private string FormatQuantiles(QuantileValue[] quantiles)
+            {
+                StringBuilder sb = new StringBuilder();
```
I believe this should be ValueStringBuilder here too.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,120 @@
+            Value1 = value1;
+            Value2 = value2;
+        }
+
```
~GetHashCode is missing.
Same on the others below.~

Ah, I see it's a partial struct and the implementation is in other files. Can you add a comment here to make this obvious?

noahfalk replied:
Sure : ) Fwiw this pattern of putting platform specific implementations in another file with partial is a pattern that pops up a good bit in the libraries code.

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.netcore.cs:
```diff
@@ -0,0 +1,65 @@
+            return _values.AsSpan();
+        }
+
+        public override int GetHashCode()
```
Could use the _instance_ methods of `HashCode`?

gfoidl commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/StringSequence.netcore.cs:
```diff
@@ -0,0 +1,56 @@
+
+    internal partial struct StringSequenceMany : IEquatable<StringSequenceMany>, IStringSequence
+    {
+        public override int GetHashCode()
```
Same here. Instance methods of `HashCode`?

noahfalk committed:
MetricEventSource cleanup

- Made some adjustments to the events
- Added a bunch of tests
- Fixed all the bugs I found with those tests
- Misc refactoring
- PR feedback

noahfalk commented:
Thanks for the nice review @gfoidl! I think almost all of it has been applied and a few parts were rendered moot from other changes.

noahfalk commented:
I got most of the changes I wanted made and rebased it on main so now it is probably more reasonable to review it.

cijothomas commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RateAggregator.cs:
```diff
@@ -0,0 +1,70 @@
+                RateStatistics? stats = default;
+                double delta = _sum - _prevSum;
+                stats = new RateStatistics(delta);
+                _prevSum = _sum;
```
if you reset _sum to zero on Collect(), then you don't need to do the delta calculation here, right?

noahfalk replied:
yep! I was doing something a little more complex before that made the 2nd sum relevant but you are right it is no longer necessary.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -9,9 +9,21 @@
     <DefineConstants Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">$(DefineConstants);ALLOW_PARTIALLY_TRUSTED_CALLERS;ENABLE_HTTP_HANDLER</DefineConstants>
-    <DefineConstants Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">$(DefineConstants);W3C_DEFAULT_ID_FORMAT</DefineConstants>
+    <DefineConstants Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">$(DefineConstants);W3C_DEFAULT_ID_FORMAT;MEMORYMARSHAL_SUPPORT;OS_SUPPORT_ATTRTIBUTES</DefineConstants>
+      <!-- To allow this library to work on V4.5 runtimes and other old platforms
```
@ericstj @ViktorHofer could you please have a look at the changes in this file?

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -9,9 +9,21 @@
+         we also have a separate complilation of this DLL that works for V4.5
+         (which is netstandard1.1).  Again we duplicate in a portable-* folder
+         to work with older NuGet clients -->
+    <PackageTargetFramework Condition="'$(TargetFramework)' == 'netstandard1.1'">netstandard1.1;portable-net45+win8+wpa81</PackageTargetFramework>
```
>netstandard1.1 [](http://example.com/codeflow?start=64&length=14)

do we ever compile ns1.1? I am not seeing it listed in the TargetFrameworks

ViktorHofer replied:
No, we don't support .NET Standard < 2.0 anymore, neither portable assets.

noahfalk replied:
I think my rebase inadvertently restored some older code in this file, let me re-scan it.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/Properties/InternalsVisibleTo.cs:
```diff
@@ -0,0 +1,6 @@
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.Diagnostics.DiagnosticSource.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22 ...
```
>InternalsVisibleTo [](http://example.com/codeflow?start=11&length=18)

I really don't like having this attribute. Is there any other way we can avoid that (using reflection for instance?

ViktorHofer replied:
If that's not possible (for whatever reason) please prefer setting the InternalsVisibleTo as an item in the project file instead.

noahfalk replied:
Why do you prefer reflection to this attribute Tarek? At present there are ~30 tests that are unit testing internal APIs. I assume with reflection we could get around it, but its not clear what value is to add all the extra code?

tarekgh replied:
In general, it creates hard dependencies. at any time, we need to use the same logic in other tests or decided for some reasons changing the test assembly names, or even the public key (in extreme cases), this will be broken. I know reflection is more involving but more flexible.

noahfalk replied:
> at any time, we need to use the same logic in other tests

I'm not sure I followed this part of the explanation. Did you mean we decide that we want to create a 2nd test assembly that has similar code in it? I'm not sure what would ever cause us to do that so it feels like an unlikely eventuality to me, but we could always add a 2nd InternalsVisibleTo attribute in such a case.

> decided for some reasons changing the test assembly names, or even the public key

If we changed the test assembly names/keys wouldn't we update the InternalsVisibleTo attribute at the same time? This one also seems like a relatively unlikely eventuality?

> I know reflection is more involving but more flexible.

I'm not sure I am understanding the full value of the flexibility yet. If the only risk we avoid is that in the future we would change/add an InternalsVisibleTo attribute that feels like the benefit is much smaller than the cost.

tarekgh replied:
> I'm not sure I followed this part of the explanation. Did you mean we decide that we want to create a 2nd test assembly that has similar code in it? I'm not sure what would ever cause us to do that so it feels like an unlikely eventuality to me, 

We run into some cases in other areas. For example, we wanted to run some standalone tests under specific environments (like with environment variables). It is testing the same area but under some different environment. 
by the way, did you try to run any test case under remote execution and findout if this work fine there?

> but we could always add a 2nd InternalsVisibleTo attribute in such a case.
> If we changed the test assembly names/keys wouldn't we update the InternalsVisibleTo attribute at the same time?

Why we'll need to touch the source assembly to just update the tests? this is exactly the point I am trying avoid :-)

> I'm not sure I am understanding the full value of the flexibility yet. If the only risk we avoid is that in the future we would change/add an InternalsVisibleTo attribute that feels like the benefit is much smaller than the cost.

The flexibility is we can change the test the way we like without ever touching the source assembly. tying the test to the source assembly is unneeded restriction IMHO.

I'll leave it to you to decide. I just explained why I don't like using this attribute :-) it shouldn't block this PR anyway.

noahfalk replied:
> I'll leave it to you to decide.

I'll stick with the InternalsVisibleTo attribute as you haven't convinced me yet, but if you change your mind and want to resume the conversation later I'm happy to do so. I appreciate the feedback even if we haven't reached agreement on it.
I think where we are leaving off is that:
- You feel needing to touch the product assembly in response to a test change is a fairly bad place to be and merits some non-trivial effort/adding complexity to avoid it.
- My take is that I don't understand why the scenarios you described would require the test assembly name to change, and that in the unlikely case the test name did need to change I don't understand why modifying the InternalsVisibleTo attribute at the same time wouldn't be straightforward.

> did you try to run any test case under remote execution and findout if this work fine there?

If you mean have I tried to use the EventSource via an out-of-proc tool rather than via EventListener, yeah I have. I have changes to dotnet-counters in progress so I am doing some ad-hoc testing. Once the dotnet-counters work completes I expect we'll have some automated testing that does validation of the E2E scenario.

stephentoub replied:
This creates other problems.  For example, the linker won't remove anything internal from the assembly, as it doesn't know what might be accessed, and it creates problems for other analyzers that simply bail on their analysis because they can't see the whole picture.  We've avoided this at almost all costs, and haven't added it for any new code; the only projects that use it today are ones that had used it prior to moving into dotnet/runtime.

noahfalk replied:
OK, the inability to do linker analysis is a problem I grok. Is the recommendation to always use reflection to test internal surface area or is there a better solution to this? As a broader issue for the .NET ecosystem I hope we aren't making people choose between easy unit testing and good linker performance.

stephentoub replied:
Some of our projects use reflection, in particular when what needs to be accessed is minimal.  Some of our projects build (a subset of) product source into the test project(s) as well; this is common in the System.Net.* libs in particular, in their UnitTest directories.

noahfalk replied:
I compiled some of the product source into the test assembly and it was pretty straightforward, thanks for the suggestion!
InternalsVisibleTo is removed.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+    [SecuritySafeCritical]
+    internal class AggregationManager
+    {
+        public const double MinCollectionTime = 0.1;
```
>. [](http://example.com/codeflow?start=49&length=1)

could you please add a comment what the unit of this value?

noahfalk replied:
I changed its name to include "Secs"

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+        private TimeSpan _collectionPeriod;
+
+
+
```
nit: remove these extra 2 empty lines

ViktorHofer commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -9,9 +9,21 @@
+         (which is netstandard1.1).  Again we duplicate in a portable-* folder
+         to work with older NuGet clients -->
+    <PackageTargetFramework Condition="'$(TargetFramework)' == 'netstandard1.1'">netstandard1.1;portable-net45+win8+wpa81</PackageTargetFramework>
+    <DefineConstants Condition="'$(TargetFramework)' == 'netstandard1.1' or '$(TargetFramework)' == 'net45'">$(DefineConstants);NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT;NO_ARRAY_EMPTY_SUPPORT</DefineConstants>
+    <DefineConstants Condition="'$(TargetFramework)' != 'netstandard1.1'">$(DefineConstants);EVENTSOURCE_ACTIVITY_SUPPORT</DefineConstants>
+    <DefineConstants Condition="'$(TargetFramework)' != 'netstandard1.1' and '$(TargetFramework)' != 'netstandard1.3'">$(DefineConstants);EVENTSOURCE_ENUMERATE_SUPPORT;IS_GENERIC_TYPE_SUPPORT</DefineConstants>
+    <ExcludeFromPackage Condition="'$(TargetFramework)' == 'netstandard2.0'">true</ExcludeFromPackage>
+    <!-- Avoid referencing NETStandard.Library on netstandard1.x since this package is part of that closure -->
+    <DisableImplicitFrameworkReferences Condition="'$(TargetFramework)' == 'netstandard1.1' or '$(TargetFramework)' == 'netstandard1.3'">true</DisableImplicitFrameworkReferences>
```
A lot of these changes are quite old and should not be brought back as we don't support netstandard1.x or net45 anymore.

noahfalk replied:
Agreed, it was unintentional

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+        }
+
+
+
```
nit: remove extra empty lines

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+
+        public AggregationManager Include(string meterName, string instrumentName)
+        {
+            Include(i => i.Meter.Name == meterName && i.Name == instrumentName);
```
>i => i.Meter.Name == meterName && i.Name == instrumentName [](http://example.com/codeflow?start=20&length=58)

just wondering, wouldn't this will create a delegate object everytim it gets called?

noahfalk replied:
yeah it does. Include only gets called when EventSource is being turned on (or reconfigured) so this happens very rarely overall.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+            if (_collectThread != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException("Start can only be called once");
```
>Start [](http://example.com/codeflow?start=53&length=5)

should we add these strings to the resources file?

noahfalk replied:
I'll follow whatever path you think is best. 
My logic for not adding them thus far is that a user should never be able to observe this exception unless our code has a bug in it. Would you like me to add them?

tarekgh replied:
Who will observe this exception message? is it only helpful for debugging? if so, then no need to store them in the resources but at least have some comments why we are doing that.

noahfalk replied:
Yeah, unless there are bugs in the implementation an end user should never see the exception. They are there just as defensive coding. In particular the ones that are checking the interval timing ensure we don't get into a tight infinite loop similar to a servicing issue we recently had to fix for EventCounter. I'll add comments.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+            // This explicitly uses a Thread and not a Task so that metrics still work
+            // even when an app is experiencing thread-pool starvation. Although we
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
```
CC @stephentoub just in case he has any comment.

stephentoub replied:
How many of these AggregationManager instances will there be in a process, and thus how many of these threads?

noahfalk replied:
At any given time there should be at most one instance of AggregationManager and one thread. MetricsEventSource disposes the old one (which joins the thread) before creating new one.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTime} sec");
+                }
+
+                DateTime startTime = DateTime.UtcNow;
```
>startTime [](http://example.com/codeflow?start=25&length=9)

would Stopwatch work better here?

noahfalk replied:
At the moment the EventSource reports a DateTime for the start of each metric collection interval. It appears difficult to preserve that behavior with Stopwatch alone because Stopwatch doesn't expose its start time. The best I see there would be to capture both DateTime.Now and start a stopwatch, then assume there neglible time passed in between the two measurements. It doesn't appear simpler but I may also be missing what you had in mind.

tarekgh replied:
if you think it is not simplifying anything then ignore this comment. I was just raising the idea.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+                {
+                    _instrumentStates.TryAdd(instrument, instrumentState);
+                    instrumentState = _instrumentStates[instrument];
+                }
```
would moving this block after the line 

instrumentState = BuildInstrumentState(instrument);

be better? or it is possible state be created in other places?

noahfalk replied:
I think there is no difference in correctness/perf so happy to swap it to the one that you felt was easier to read.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,312 @@
+        internal InstrumentState? BuildInstrumentState(Instrument instrument)
+        {
+            Func<Aggregator>? createAggregatorFunc = GetAggregatorFactory(instrument);
+            if (createAggregatorFunc == null)
```
>createAggregatorFunc == null [](http://example.com/codeflow?start=16&length=28)

nit: do you prefer this syntax or you prefer using 
if (x is null) and if (x is not null)?

noahfalk replied:
My personal preference is x == null but I'm happy to do it differently if there is a general convention or consensus that 'is null' is the better one?

tarekgh replied:
It is not a big deal. I used the other style in the Metrics code. this is why I asked about it :-)

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,468 @@
+            where TAggregator : Aggregator
+        {
+            LabelInstruction[] instructions = Compile(labels);
+            Array.Sort(instructions, (LabelInstruction a, LabelInstruction b) => a.LabelName.CompareTo(b.LabelName));
```
>CompareTo [](http://example.com/codeflow?start=93&length=9)

please use CompareOrdinal here.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,468 @@
+        }
+    }
+
+    internal struct LabelInstruction
```
>LabelInstruction [](http://example.com/codeflow?start=20&length=16)

make it readonly and provide a constructor? this may help a little with the perf I guess.

stephentoub replied:
```suggestion
    internal readonly struct LabelInstruction
```

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,356 @@
+                    if (command.Arguments!.TryGetValue("RefreshInterval", out string? refreshInterval))
+                    {
+                        Logger.Message("RefreshInterval filter argument received: " + refreshInterval);
+                        if (!double.TryParse(refreshInterval, out refreshIntervalSecs))
```
>refreshIntervalSecs [](http://example.com/codeflow?start=66&length=19)

should we check if this value less than MinCollectionTime?

noahfalk replied:
I added an assert that should start firing if someone accidentally makes default < MinCollectionTime. Is that what you meant here?

tarekgh replied:
yes. I am just wondering instead of the assert you rest the value to the min value?

noahfalk replied:
If a dev changes either the minimum or the default and it causes the default to no longer be valid I'd like them to see their mistake quickly and address it. I worry that coercing the value would obscure the issue and leave the dev unaware that their change had unintended results.

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,356 @@
+namespace System.Diagnostics.Metrics
+{
+    [EventSource(Name = "System.Diagnostics.Metrics")]
+    internal class MetricsEventSource : EventSource
```
>MetricsEventSource [](http://example.com/codeflow?start=19&length=18)

could you please add some comment here regarding the spec syntax and listing some examples? I didn't review all files yet so far I didn't see if you already have it somewhere else.

noahfalk replied:
Nope you are accurate, it hadn't been documented in the comments yet : )

tarekgh commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,356 @@
+                }
+                else
+                {
+                    string meterName = text.Substring(0, slashIdx).Trim();
```
>meterName [](http://example.com/codeflow?start=27&length=9)

if I pass empty string does it mean listening to all meters/instruments? so we only support specific meters and instruments? 

I am asking to know if we can listen to more than one meter and instrumts.

noahfalk replied:
Empty string would only match a meter whose name was empty string. It is possible to listen to multiple Meters/Instruments by passing a semicolon separated list as the Metrics argument. This test case shows an example: https://github.com/dotnet/runtime/pull/54333/files/f871425333a4f5a37e6a952f38f4943768db7d76#diff-8797eae6d3e644ed047f952e10d4240fe18175f754586666823b555f46968672R262

tarekgh replied:
do we support wild cards?

noahfalk replied:
We support using the name of the meter to include all instruments on that meter, but otherwise no. This comment is the full description of what is currently supported:
https://github.com/dotnet/runtime/pull/54333/files#diff-072334d2e8cacd4f335b7f48f0da8249a5619b6237e9ed311b029801145e8022R31-R34

noahfalk committed:
PR feedback

noahfalk committed:
More PR feedback

noahfalk committed:
- Add some additional error reporting

- Disable tests on browser where the feature isn't supported

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+{
+#if OS_SUPPORT_ATTRTIBUTES
+    [UnsupportedOSPlatform("browser")]
+#endif
```
Instead of ifdef'ing these, can we add internal versions that are compiled in for platforms that don't have them?  That's what we do for nullable attributes, for example:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs
https://github.com/dotnet/runtime/blob/f399076cb971fcfdab397d8b3786f321acb4b4d5/src/libraries/Directory.Build.targets#L236-L239
and in general it's the plan the C# compiler is on, e.g. if it finds the attributes it needs, it uses them, otherwise it emits them as internal into the assembly.
cc: @buyaa-n

noahfalk replied:
It appears there is an opt-in property designed for the attributes already and I switched to use it: https://github.com/dotnet/runtime/blob/f399076cb971fcfdab397d8b3786f321acb4b4d5/src/libraries/Directory.Build.targets#L255-L257

It won't cover OperatingSystem.IsBrowser() API used here: https://github.com/dotnet/runtime/pull/54333/files#diff-072334d2e8cacd4f335b7f48f0da8249a5619b6237e9ed311b029801145e8022R180
and so far the [precedent](https://source.dot.net/#System.Runtime.Caching/System/Runtime/Caching/HostFileChangeMonitor.cs,101) I saw is to ifdef it in place. Let me know if you prefer establishing a different pattern.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+    internal class AggregationManager
+    {
+        public const double MinCollectionTimeSecs = 0.1;
+        private static readonly QuantileAggregation DefaultHistogramConfig = new QuantileAggregation(new double[] { 0.50, 0.95, 0.99 });
```
Nit: s_defaultHistogramConfig

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        private List<Predicate<Instrument>> _instrumentConfigFuncs = new();
+        private TimeSpan _collectionPeriod;
+
+        private ConcurrentDictionary<Instrument, InstrumentState> _instrumentStates = new();
```
```suggestion
        private readonly ConcurrentDictionary<Instrument, InstrumentState> _instrumentStates = new();
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        private TimeSpan _collectionPeriod;
+
+        private ConcurrentDictionary<Instrument, InstrumentState> _instrumentStates = new();
+        private CancellationTokenSource _cts = new();
```
```suggestion
        private readonly CancellationTokenSource _cts = new();
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        private Action<Instrument> _endInstrumentMeasurements;
+        private Action<Instrument> _instrumentPublished;
+        private Action _initialInstrumentEnumerationComplete;
+        private Action<Exception> _collectionError;
```
Seems like most of these could be readonly as well?

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        public AggregationManager Include(string meterName)
+        {
+            Include(i => i.Meter.Name == meterName);
+            return this;
```
Why do these Include methods return `this` if none of the call sites pay attention to the return value?

noahfalk replied:
Originally there was some code that was making calls in a fluent style. That code is gone now and 'return this' is a remnant that I forgot about. We can make it void.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+    [UnsupportedOSPlatform("browser")]
+#endif
+    [SecuritySafeCritical]
+    internal class AggregationManager
```
```suggestion
    internal sealed class AggregationManager
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            {
+                // this is purely for defensive programming and it should be impossible
+                // for a developer to observe this exception unless this library has a
+                // bug in it. Because of that we didn't bother to localize the error message.
```
Should we just assert it instead then?
```C#
Debug.Assert(_collectThread is null && !_cts.IsCancellationRequested);
```
?

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            // even when an app is experiencing thread-pool starvation. Although we
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
+            _collectThread = new Thread(() => CollectWorker(_cts.Token));
```
If we are going to use a dedicated thread, it should be IsBackground = true, and it should be given a name to aid in debugging.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
+            _collectThread = new Thread(() => CollectWorker(_cts.Token));
+            _collectThread.Start();
```
Does this thread invoke user code?  If yes, this is good.  If no, it should use UnsafeStart on netcoreappcurrent.

noahfalk replied:
Yes, it runs user provided callbacks at times.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+                    // this is purely for defensive programming and it should be impossible
+                    // for a developer to observe this exception unless this library has a
+                    // bug in it. Because of that we didn't bother to localize the error message.
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTimeSecs} sec");
```
Same comment.  If the only way to hit this is because we have a bug, it should be an assert.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+
+                    // pause until the interval is complete
+                    TimeSpan delayTime = nextIntervalStartTime - now;
+                    cancelToken.WaitHandle.WaitOne(delayTime);
```
I think this may be the only use of CancellationToken.WaitHandle in all of our product source. :-)

noahfalk replied:
I was hoping for Thread.Sleep(timeout, cancelToken) originally, but this also did what I needed it to do.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            }
+            catch (Exception e)
+            {
+                _collectionError(e);
```
What happens when this fails?  Is there a mechanism for recovery, since Start can only be used once?

noahfalk replied:
The recovery mechanism at the moment is to Dispose() the manager and create a new one. We might need a more localized handler to make Exceptions in the user callbacks less problematic as well but I haven't scrutinized that aspect just yet.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+                    cancelToken.WaitHandle.WaitOne(delayTime);
+
+                    // don't do collection if timer may not have run to completion
+                    if (cancelToken.IsCancellationRequested)
```
You can just check the return value of WaitOne.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+
+        private void RemoveInstrumentState(Instrument instrument, InstrumentState state)
+        {
+            _instrumentStates.TryRemove(instrument, out InstrumentState? _state);
```
Nit:
```suggestion
            _instrumentStates.TryRemove(instrument, out _);
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
```
Nit: var => whatever type this is

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        {
+            if (!_instrumentStates.TryGetValue(instrument, out InstrumentState? instrumentState))
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
```
```suggestion
                lock (this) // protect _instrumentConfigFuncs list
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+                            instrumentState = BuildInstrumentState(instrument);
+                            if (instrumentState != null)
+                            {
+                                _instrumentStates.TryAdd(instrument, instrumentState);
+                                instrumentState = _instrumentStates[instrument];
```
```suggestion
                                if (!_instrumentStates.TryAdd(instrument, instrumentState))
                                {
                                    instrumentState = _instrumentStates[instrument];
                                }
```

noahfalk replied:
Some technicalities elsewhere were probably saving me, but I'd rather do it the way you suggest and not play with the fire : )

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            {
+                return null;
+            }
+            Type aggregatorType = createAggregatorFunc.GetType().GenericTypeArguments[0];
```
Is this type guaranteed to be generic?

stephentoub replied:
(Nevermind, I see GetAggregatorFactory below guarantees it.)

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        {
+            Type type = instrument.GetType();
+            Type? genericDefType = null;
+#if IS_GENERIC_TYPE_SUPPORT
```
Why is this #if needed?

noahfalk replied:
While I was doing development I think a concurrent change came in that removed the support for one of the old targets this library supported (netstandard1.1 perhaps?). The rebase eliminated the define in the project but I had forgotten to remove the usage in the code.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+                kv.Value.Collect(kv.Key, (LabeledAggregationStatistics labeledAggStats) =>
+                {
+                    _collectMeasurement(kv.Key, labeledAggStats);
+                });
```
How frequently does this code run / how much do we care about overheads?  This, for example, is allocating a new delegate and a new closure for every pair in _instrumentStates.

noahfalk replied:
In a typical scenario Collect() runs between once per second and once per 10 minutes. In the limit it could run 10 times/sec. Typical scenario probably has 10-100 items in this loop. 

Thus far I paid very little care to allocations in this path with the expectation the scenario impact would be neglible.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Aggregator.cs:
```diff
@@ -0,0 +1,52 @@
+
+    internal interface IAggregationStatistics { }
+
+    internal struct QuantileValue
```
```suggestion
    internal readonly struct QuantileValue
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Aggregator.cs:
```diff
@@ -0,0 +1,52 @@
+        public double Value { get; }
+    }
+
+    internal class HistogramStatistics : IAggregationStatistics
```
```suggestion
    internal sealed class HistogramStatistics : IAggregationStatistics
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Aggregator.cs:
```diff
@@ -0,0 +1,52 @@
+        public QuantileValue[] Quantiles { get; }
+    }
+
+    internal class LabeledAggregationStatistics
```
```suggestion
    internal sealed class LabeledAggregationStatistics
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+        // MultiSizeLabelNameDictionary<TAggregator> - this is used when we need to store more than one of the above union items
+        private volatile object? _stateUnion;
+        private volatile AggregatorLookupFunc<TAggregator>? _cachedLookupFunc;
+        private Func<TAggregator> _createAggregatorFunc;
```
```suggestion
        private readonly Func<TAggregator> _createAggregatorFunc;
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+            AggregatorLookupFunc<TAggregator>? lookupFunc = _cachedLookupFunc;
+            if (lookupFunc != null)
+            {
+                if (lookupFunc(labels, out TAggregator? aggregator)) return aggregator!;
```
This `!` shouldn't be necessary if you change the definition of the `AggregatorLookupFunc` delegate to have a `[NotNullWhen(true)]` attribute on the out argument: `[NotNullWhen(true)] out TAggregator? aggregator`.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+            else if (stateUnion is MultiSizeLabelNameDictionary<TAggregator> aggsMultiSize)
+            {
+                aggsMultiSize.Collect(visitFunc);
+            }
```
Beauty is in the eye of the beholder, but if you want, you can rewrite this as a pattern matching switch:
```C#
switch (_stateUnion)
{
    case TAggregator agg:
        IAggregationStatistics stats = agg.Collect();
        visitFunc(new LabeledAggregationStatistics(stats));
        break;

    case FixedSizeLabelNameDictionary<StringSequence1, ObjectSequence1, TAggregator> aggs1:
        aggs1.Collect(visitFunc);
        break;

    case FixedSizeLabelNameDictionary<StringSequence2, ObjectSequence2, TAggregator> aggs2:
        aggs2.Collect(visitFunc);
        break;

    ...
}
```
I personally prefer the switch style, but up to you.

noahfalk replied:
I'm happy to match team preference : )

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+                if (state == null)
+                {
+                    TAggregator newState = _createAggregatorFunc();
+                    Interlocked.CompareExchange(ref _stateUnion, newState, null);
+                    continue;
```
```suggestion
                    if (Interlocked.CompareExchange(ref _stateUnion, newState, null) is null)
                    {
                        return newState;
                    }
                    continue;
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+                else
+                {
+                    MultiSizeLabelNameDictionary<TAggregator> newState = new(state);
+                    Interlocked.CompareExchange(ref _stateUnion, newState, state);
```
```suggestion
                    if (Interlocked.CompareExchange(ref _stateUnion, newState, state) == state)
                    {
                        return multiSizeState.GetNoLabelAggregator(_createAggregatorFunc);
                    }
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+                if (state == null)
+                {
+                    FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator> newState = new();
+                    Interlocked.CompareExchange(ref _stateUnion, newState, null);
```
Same suggestions here.  Races should be rare (yes?) so we can optimize for the success case and avoid the extra loop around, type checks, and memory barrier (on arm).

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+            TStringSequence? seq = default;
+            if (seq is StringSequence1)
+            {
+                if (Label1 == null)
```
Some of this code is using default, some is using null.  It'd be good to be consistent.  Personally I'd prefer if all of the reference types used null.

noahfalk replied:
I swapped default to null where allowable

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+                }
+                return (FixedSizeLabelNameDictionary<TStringSequence, TObjectSequence, TAggregator>)(object)LabelMany;
+            }
+            // we should never get here unless this library has a bug
```
Same assert comments as earlier.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+            SourceIndex = sourceIndex;
+            LabelName = labelName;
+        }
+        public readonly int SourceIndex;
+        public readonly string LabelName;
```
Elsewhere with similar structs you had these as auto-props.  As this is all internal, it doesn't really matter which you use, but it'd be nice to be consistent.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+        private LabelInstruction[] _instructions;
+        private ConcurrentDictionary<TObjectSequence, TAggregator> _valuesDict;
+        private Func<TObjectSequence, TAggregator> _createAggregator;
+
+
```
```suggestion

```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+
+namespace System.Diagnostics.Metrics
+{
+    internal class QuantileAggregation
```
```suggestion
    internal sealed class QuantileAggregation
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+    // PERF Note: This histogram has a fast Update() but the _counters array has a sizable of memory footprint (32KB+ on 64 bit)
+    // It is probably well suited for tracking 10s or maybe 100s of histograms but if we wanted to go higher
+    // we probably want to trade a little more CPU cost in Update() + code complexity for more precise array memory growth.
+    internal class ExponentialHistogramAggregator : Aggregator
```
```suggestion
    internal sealed class ExponentialHistogramAggregator : Aggregator
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs:
```diff
@@ -0,0 +1,472 @@
+    }
+
+    [System.Security.SecurityCritical] // using SecurityCritical type ReadOnlySpan
+    internal class LabelInstructionInterpretter<TObjectSequence, TAggregator>
```
```suggestion
    internal sealed class LabelInstructionInterpretter<TObjectSequence, TAggregator>
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+        private readonly int _mantissaMask;
+        private readonly int _mantissaShift;
+
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+                count = _count;
+                _counters = new int[ExponentArraySize][];
+                _count = 0;
+            }
+            QuantileValue[] quantiles = new QuantileValue[_config.Quantiles.Length];
```
```suggestion
            }

            QuantileValue[] quantiles = new QuantileValue[_config.Quantiles.Length];
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+            }
+
+            Debug.Assert(count == 0);
+#if NO_ARRAY_EMPTY_SUPPORT
```
What platform do we target that doesn't have Array.Empty?

noahfalk replied:
It was one of the old ones (net45 or netstandard1.1?). I think it was recently eliminated as a target so after the rebase this became dead code.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+#endif
+        }
+
+        private int GetInvalidCount(int[][] counters)
+        {
+            int[] positiveInfAndNan = counters[ExponentArraySize / 2 - 1];
+            int[] negativeInfAndNan = counters[ExponentArraySize - 1];
+            int count = 0;
+            if (positiveInfAndNan != null)
```
I'm confused by the nullability annotations here.  If positiveInfAndNan could be null, seems we're missing some question marks in various places.

noahfalk replied:
Fixed. It should have been int[]?[]

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs:
```diff
@@ -0,0 +1,216 @@
+        private IEnumerable<Bucket> IterateBuckets(int[][] counters)
+        {
+            // iterate over the negative exponent buckets
+            const int lowestNegativeOffset = ExponentArraySize / 2;
```
Nit: lowestNegativeOffset => LowestNegativeOffset

noahfalk replied:
just curious, do you expect it would be hard to create analyzers to enforce some of the naming and capitalization suggestions? I appreciate you taking the time to review it all but I also feel bummed that you would need to.

noahfalk replied:
(edit has been applied)

stephentoub replied:
They exist.  We're just not yet consistent enough in some of our existing code to enable it, and there's no great baselining mechanism.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/LastValueAggregator.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace System.Diagnostics.Metrics
+{
+    internal class LastValue : Aggregator
```
```suggestion
    internal sealed class LastValue : Aggregator
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/LastValueAggregator.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    internal class LastValueStatistics : IAggregationStatistics
```
```suggestion
    internal sealed class LastValueStatistics : IAggregationStatistics
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+{
+    /// <summary>
+    /// This EventSource is intended to let out-of-process tools (such as dotnet-counters) do
+    /// ad-hoc monitoring for the new Instrument APIs. This source only supports one listener
+    /// at a time. Each new listener will overwrite the configuration about which metrics
```
Isn't that broken?  This is saying if I add my own EventListener and happen to enable the events from this event source, things break?

noahfalk replied:
it is certainly a limitation in the scenario, but it shouldn't render it useless. The main intended use case for this is ad-hoc collection tools and we'd tell people don't try to run multiple of these tools concurrently. If they do it anyways we can identify that happened and give an error message. For production/always-on monitoring scenarios they would be better off using a library like OpenTelemetry to export the data continuously. I also have some thoughts how we can eliminate this limitation in .NET 7.

josalem replied:
Rather than having multiple session overwrite the configuration in 6, could we keep track of a union of requested metrics?

All sessions would observe additions to the metric collection and increases in the frequency.

e.g.,

Session A enables Metric1;Metric2 @ 10s intervals.
Session B enables Metric1;Metric3 @  5s intervals.

then both sessions would see Metric1;Metric2;Metric3 @ 5s intervals.

Session B disables.

Session A continues to see Metric1;Metric2;Metric3 @ 5s intervals.

Then in 7 we could attempt to make changes unique to a session which would require more bookkeeping like your comment says.

noahfalk replied:
I think its possible, but it gets a little messy and I am not sure I have the time to work through the details. For example the requested intervals may not be multiples of one another, it requires the tools to re-filter, and it opens the door to having large numbers of metrics reported that won't stop reporting even after the tool disconnects.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+    ///   in the 'System.Runtime' meter.
+    /// </summary>
+    [EventSource(Name = "System.Diagnostics.Metrics")]
+    internal class MetricsEventSource : EventSource
+    {
```
```suggestion
    internal sealed class MetricsEventSource : EventSource
    {
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+    [EventSource(Name = "System.Diagnostics.Metrics")]
+    internal class MetricsEventSource : EventSource
+    {
+        public static readonly MetricsEventSource Logger = new();
```
Nit: Isn't this typically called Log rather than Logger?

noahfalk replied:
Often it is. This one was copied from DiagnosticSourceEventSource which happened call it 'Logger'. I just changed both of them to Log to eliminate the discrepancy.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+        // Methods that are declared explicitly can use the [NonEvent] attribute to opt-out but
+        // lambdas can't. Putting all the command handling logic in this nested class
+        // is a simpler way to opt everything out in bulk.
+        private class CommandHandler
```
```suggestion
        private sealed class CommandHandler
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+                        {
+                            _aggregationManager.Dispose();
+                            _aggregationManager = null;
+                            Logger.Message($"Previous session with id {_sessionId} is stopped");
```
If we're here, are these events guaranteed to be enabled?  If not, we're potentially doing formatting work that'll just be throwaway.

noahfalk replied:
It is not guaranteed the events are enabled, but these code paths should run very rarely and I wouldn't suggest writing extra code to optimize it.

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+                        double refreshIntervalSecs = defaultIntervalSecs;
+                        if (command.Arguments!.TryGetValue("RefreshInterval", out string? refreshInterval))
+                        {
+                            Logger.Message("RefreshInterval filter argument received: " + refreshInterval);
```
For consistency with the other tracing:
```suggestion
                            Logger.Message($"RefreshInterval filter argument received: {refreshInterval}");
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+                return false;
+            }
+
+            private static char[] s_instrumentSeperators = new char[] { '\r', '\n', ',', ';' };
```
```suggestion
            private static readonly char[] s_instrumentSeperators = new char[] { '\r', '\n', ',', ';' };
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+                {
+                    if (!MetricSpec.TryParse(specString, out MetricSpec spec))
+                    {
+                        Logger.Message("Failed to parse metric spec: " + specString);
```
```suggestion
                        Logger.Message($"Failed to parse metric spec: {specString}");
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+                    }
+                    else
+                    {
+                        Logger.Message("Parsed metric: " + spec.ToString());
```
```suggestion
                        Logger.Message("Parsed metric: {spec}");
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+
+        private class MetricSpec
+        {
+            private static char s_meterInstrumentSeparator = '\\';
```
```suggestion
            private const char MeterInstrumentSeparator = '\\';
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -0,0 +1,395 @@
+            }
+
+            public override string ToString() => MeterName +
+                (InstrumentName != null ? s_meterInstrumentSeparator + InstrumentName : "");
```
```suggestion
                (InstrumentName != null ? MeterInstrumentSeparator + InstrumentName : "");
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+            Value1 = value1;
+        }
+
+        public override int GetHashCode() => Value1?.GetHashCode() ?? 0;
```
Is Value1 ever written outside of the ctor?  If no, it should be readonly.  If yes, does any code depend on this GetHashCode being stable, e.g. are these ever put into a dictionary?

noahfalk replied:
Is Value1 ever written outside of the ctor? Nope

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+
+        public bool Equals(ObjectSequence1 other)
+        {
+            return (Value1 == null && other.Value1 == null) || (Value1 != null && Value1.Equals(other.Value1));
```
```suggestion
            return Value1 is null ? other.Value1 is null : Value1.Equals(other.Value1);
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+        //GetHashCode() is in the platform specific files
+        public override bool Equals(object? obj)
+        {
+            return obj is ObjectSequence1 && Equals((ObjectSequence1)obj);
```
```suggestion
            return obj is ObjectSequence1 os1 && Equals(os1);
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    internal partial struct ObjectSequence2 : IEquatable<ObjectSequence2>, IObjectSequence
```
Same comments as on OS1

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    internal partial struct ObjectSequence3 : IEquatable<ObjectSequence3>, IObjectSequence
```
Same comments as on OS1

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+
+    internal partial struct ObjectSequenceMany : IEquatable<ObjectSequenceMany>, IObjectSequence
+    {
+        private object?[] _values;
```
```suggestion
        private readonly object?[] _values;
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+            }
+            for (int i = 0; i < _values.Length; i++)
+            {
+                if (_values[i] == null)
+                {
+                    if (other._values[i] == null)
+                    {
+                        continue;
+                    }
+                    return false;
+                }
+                if (!_values[i]!.Equals(other._values[i]))
+                {
+                    return false;
+                }
```
```suggestion
                object? value = _values[i], otherValue = other._values[i];
                if (value is null)
                {
                    if (otherValue is not null)
                    {
                        return false;
                    }
                }
                else if (!value.Equals(otherValue))
                {
                    return false;
                }
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs:
```diff
@@ -0,0 +1,122 @@
+        //GetHashCode() is in the platform specific files
+        public override bool Equals(object? obj)
+        {
+            return obj is ObjectSequenceMany && Equals((ObjectSequenceMany)obj);
```
```suggestion
            return obj is ObjectSequenceMany osm && Equals(osm);
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.netcore.cs:
```diff
@@ -0,0 +1,60 @@
+
+    internal partial struct ObjectSequenceMany : IEquatable<ObjectSequenceMany>, IObjectSequence
+    {
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RateAggregator.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace System.Diagnostics.Metrics
+{
+    internal class RateSumAggregator : Aggregator
```
```suggestion
    internal sealed class RateSumAggregator : Aggregator
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RateAggregator.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    internal class RateAggregator : Aggregator
```
```suggestion
    internal sealed class RateAggregator : Aggregator
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/RateAggregator.cs:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    internal class RateStatistics : IAggregationStatistics
```
```suggestion
    internal sealed class RateStatistics : IAggregationStatistics
```

stephentoub commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/StringSequence.cs:
```diff
@@ -0,0 +1,116 @@
+
+namespace System.Diagnostics.Metrics
+{
+    internal partial struct StringSequence1 : IEquatable<StringSequence1>, IStringSequence
```
Same comments on all of these as on the ObjectSequenceN types.

josalem commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/LastValueAggregator.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public override IAggregationStatistics Collect()
+        {
+            lock (this)
+            {
+                LastValueStatistics stats = new LastValueStatistics(_lastValue);
+                _lastValue = null;
+                return stats;
+            }
```
Does `Update` need this lock too? If it isn't locking also, I'm not sure if we need this lock here. We could swap for an interlocked exchange or something.

noahfalk replied:
Yeah, you could do this one with an InterlockedExchange instead of the lock. I left it as lock because it was simple and I don't anticipate this code is on the hot path. The hot paths are InstrumentState.Update(), AggregatorStore.GetAggregator(), RateSumAggregator.Update() and Histogram.Update(). Everything else is unlikely to go above a few hundred invocations/sec.

stephentoub replied:
> Everything else is unlikely to go above a few hundred invocations/sec.

Which methods might be invoked several hundred times a second?  Ones that might allocate?

noahfalk replied:
> Which methods might be invoked several hundred times a second? Ones that might allocate?

Yeah, if someone requested to collect metrics once per second and enabled a few hundred metrics then the InstrumentState.Collect() path would be running a few hundred times per second. InstrumentState.Collect() invokes AggregatorStore.Collect() which in turn invokes Aggregator.Collect(). There are a few allocations down that path.

None of that is going to happen automatically, an engineer needed to run a diagnostic tool to turn it on, they needed to specify they wanted collections every second, and they needed to specify a bunch of metrics they wanted collected.

noahfalk committed:
PR feedback and cleanup

noahfalk commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+            else
+            {
+                return null;
+            }
```
I don't think it works? `case typeof(typename)` gives an error that the value isn't constant. Type matching on instrument would require I know the exact type I think but I only know a partial type

noahfalk commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs:
```diff
@@ -0,0 +1,314 @@
+        {
+            _listener.RecordObservableInstruments();
+
+            foreach (KeyValuePair<Instrument, InstrumentState> kv in _instrumentStates)
```
that's cool : )

josalem replied:
Not worth it if you need to write a method for it. I thought it would work for free due to [`KeyValuePair<Tk,Tv>.Deconstruct(out Tk key, out Tv value)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.keyvaluepair-2.deconstruct?view=net-5.0).

noahfalk replied:
This library builds for some older .NET Framework versions where that API wasn't yet available.

noahfalk committed:
- PR feedback

- bugfix to use invariant culture
- support time series and histogram limits

noahfalk commented:
Thanks for all the review feedback everyone! I think everything has been addressed at this point. There are a few comments above where someone asked a question and I answered it that I didn't resolve so that folks could see the answers to their questions.

If there is any remaining feedback please let me know, otherwise I plan to hit the merge button tomorrow.

Also as a heads up I've got at least one more change I am thinking to make in a separate PR to improve how the EventSource handles exceptions that are thrown from user provided callbacks.

AndyAyersMS commented:
This bit of code two consecutive fields of an object as a span of the two fields (similarly for related types).

I suspect this is treading on dangerous territory and may lead to incorrect optimizations by the jit, especially if later on there is code that intermingles references from the `AsSpan` and the individual fields. I don't see that happening anywhere but generally speaking the jit will not properly handle the types of aliasing this can introduce.

https://github.com/dotnet/runtime/blob/5705c9863578baa5b7bd3cbb06ab7e524f541b5a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.cs#L32-L37

https://github.com/dotnet/runtime/blob/5705c9863578baa5b7bd3cbb06ab7e524f541b5a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.netcore.cs#L22-L30
